private/with-npm-packages/README.md

```diff
@@ -0,0 +1,116 @@
+| 8  |    1000   |       5000       |       1000       |    p#A    | 1569  |
+| 9  |    1000   |       5000       |       1000       |    p#B    | 1528  |
+| 10 |    1000   |       5000       |       5000       |    app    | 4547  |
+| 11 |    1000   |       5000       |       5000       |    p#A    | 3365  |
+| 12 |    1000   |       5000       |       5000       |    p#B    | 3127  |
+
+> Note: The times showed in the table above correspond to the times on code change with meteor cached.
+
+### Analysis
+
+After the previous results on the experiment we can clearly see that:
+
+- Changing on package is faster than doing it on app level. The reason for that is that with babel compilation we have gained paralelism in package compilation and transpilation on modules on demand. Meteor only has to put together the app modules and NPM dependencies. (See `3` and `4`, and `10`, `11` and `12`).
+
+In comparision with [the previous experiment](../with-meteor-packages/README.md#Analysis), we can clearly see that:
+
+- Meteor apps are better optimized than meteor packages, but meteor builder is a bit outdated in terms of project construction. For the moment, we can only ensure that having NPM packages and an exernal way to do ES6 transpilation for those, is better than having Meteor to handle in its own means.
+
+As always, during the experiment I also noted that:
+
+- Having large amount of files and links between them is critical on app level and makes the initial load of the builder (timings without the cache) to get so much time to rebuild (many many minutes! insane!).
+
+### Improvements
+
+Meteor builder seems to not be optimized on projects with large amount of files. For that I really think that adopting this approach should consider the next improvements:
+
+- Describe the less modules and files on app level. The application level should only have modules that depends on build-in Meteor stuff (like subscriptions, collections and etc) and the startup procedures. Maybe this can be enhanced further by describing the modules that depend on Meteor in NPM packages but as module factories, and use the app level to instantiate those modules by passing the needed Meteor context.
+
+- Adopt Webpack for package compilation is not possible apparently, as Webpack not only transpile ES6, also perform Meteor builder work, that is joining all modules together. But in our package level we want to have modules and submodules defined, not only single entry points generated by Webpack build. That is why babel transpilation is our best bet so far.
+
+- Use [gulpjs](https://gulpjs.com/) task runner to improve the definition of babel transpilation to all packages within `/packages` as supports wildcard definitions.
+
+- Include `babel-transform-imports` plugin to translate our index imports to specific ones, provided that [I implemented regex support for it](https://bitbucket.org/amctheatres/babel-transform-imports).
```

Comment:
I can't see any benefit yet with indexes, just cause problems with a lot of files and worst they are always called "index.js" which is pretty bad for who works with "Open file" command from IDEs.

We need to discuss that and decide it, with the IDE working fine we don't need to write any import manually, we don't even need to look to imports sections of the file.

Reply:
> I can't see any benefit yet with indexes, just cause problems with a lot of files

It is not the amount of files the issue, at least according to the results I have got in the experiments, it is a matter of the links between them. For instance,

The next statement imports the whole index main module from the package, which imports all other sample modules.
``` javascript
import { x1 as x1B } from 'meteor/b';
```
Having 5000 files indexed results in generate a bundle including all of them, which adds more time, affecting the "BuildApp" phase.

By using the specific import, we save Meteor to reach other sample modules, just `x1` as you can see in the next statement.

``` javascript
import x1 from 'meteor/b/x1';
```

The `babel-transform-imports` is an attempt to save time on the `BuildApp` phase. Only by using the NPM packages, we have saved significatively the `prepareProjectToBuild` phase due to package change. As an example, see the total time recorded in `4`, `6`, `8`, `9`, `11` and `12` experiments on both [`with-meteor-packages`](https://github.com/filipenevola/build-performance-9642/tree/master/private/with-meteor-packages#experiment) and [`with-npm-packages`](https://github.com/Gywem/build-performance-9642/blob/feature/experiment-npm-packages/private/with-npm-packages/README.md#experiment). Once I perform the babel import transformation experiment, I expect to reduce app code change scenarios, like `1`, `2`, `3`, `5`, `7` and `10`. Let see how it will go. Battle on! ⚔️

> and worst they are always called "index.js" which is pretty bad for who works with "Open file" command from IDEs.

I've use a setting in vscode to overcome such issue `"workbench.editor.labelFormat": "short",`. I remember there was other vscode plugin that even gets rid of the index.js name and put the folder context instead.

> We need to discuss that and decide it, with the IDE working fine we don't need to write any import manually, we don't even need to look to imports sections of the file.

That is right. Not sure the best approach. I usually tend to think that with index all import blocks are neater and cleaner, the IDE can help to obscure that, but on code review at github for example, the look would be still bad.
In comparison, the index files affect on establishing a code convention and use a plugin to do automatically, you keep the code clean and the specific imports are obscured by the babel transpilation process.

Reply:
I think we'll need to decide it by voting hehe

but if I make one feature without any index.js file and you need to work on that code I believe you'll like it more than indexesssssss :wink:

Reply:
> but if I make one feature without any index.js file and you need to work on that code I believe you'll like it more than indexesssssss 😉

I was planning to contribute in the `babel-transform-imports` project an option to only apply the specific import transformation if the file actually exists. I don't think that just this check introduces a lot of time for the transformation.

Other option is just apply the configuration in some folder names, like `components`, `containers`, `collections`, etc, and omit `constants` (we don't want separate by simple contants files). All is about to specifying conventions. 😄